source/common/api/api_impl.cc

```diff
@@ -11,13 +11,18 @@ namespace Api {
 
 Impl::Impl(Thread::ThreadFactory& thread_factory, Stats::Store& store,
            Event::TimeSystem& time_system, Filesystem::Instance& file_system,
-           Random::RandomGenerator& random_generator, const ProcessContextOptRef& process_context)
+           Random::RandomGenerator& random_generator, const ProcessContextOptRef& process_context,
+           Buffer::WatermarkFactorySharedPtr watermark_factory)
     : thread_factory_(thread_factory), store_(store), time_system_(time_system),
       file_system_(file_system), random_generator_(random_generator),
-      process_context_(process_context) {}
+      process_context_(process_context), watermark_factory_(std::move(watermark_factory)) {}
 
 Event::DispatcherPtr Impl::allocateDispatcher(const std::string& name) {
-  return std::make_unique<Event::DispatcherImpl>(name, *this, time_system_);
+  if (watermark_factory_) {
+    return std::make_unique<Event::DispatcherImpl>(name, watermark_factory_, *this, time_system_);
+  } else {
+    return std::make_unique<Event::DispatcherImpl>(name, *this, time_system_);
+  }
```

Comment:
nit: seems like we could collapse this into a single constructor?

Reply:
Changed the argument order and collapsed to a single DipatcherImpl contructor with an optional watermark_factory argument at the end.